account-kit/smart-contracts/src/ma-v2/account/account.ts

```diff
@@ -0,0 +1,158 @@
+import { standardExecutor } from "../../msca/account/standardExecutor.js";
+import { singleSignerMessageSigner } from "../modules/single-signer-validation/signer.js";
+
+export const DEFAULT_OWNER_ENTITY_ID = 0;
+
+export type SMAV2Account<
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+> = SmartContractAccountWithSigner<"SMAV2Account", TSigner, "0.7.0">;
+
+export type CreateSMAV2AccountParams<
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner,
+  TEntryPointVersion extends "0.7.0" = "0.7.0"
+> = Pick<
+  ToSmartContractAccountParams<
+    "SMAV2Account",
+    TTransport,
+    Chain,
+    TEntryPointVersion
+  >,
+  "transport" | "chain"
+> & {
+  signer: TSigner;
+  salt?: bigint;
+  factoryAddress?: Address;
+  initCode?: Hex;
+  initialOwner?: Address;
+  accountAddress?: Address;
+  entryPoint?: EntryPointDef<TEntryPointVersion, Chain>;
+  isGlobalValidation?: boolean;
+  entityId?: bigint;
+};
+
+export async function createSMAV2Account<
+  TTransport extends Transport = Transport,
+  TSigner extends SmartAccountSigner = SmartAccountSigner
+>(
+  config: CreateSMAV2AccountParams<TTransport, TSigner>
+): Promise<SMAV2Account<TSigner>>;
+
+export async function createSMAV2Account(
+  config: CreateSMAV2AccountParams
+): Promise<SMAV2Account> {
+  const {
+    transport,
+    chain,
+    signer,
+    salt = 0n,
+    factoryAddress = addresses.accountFactory,
+    initCode,
+    initialOwner,
+    accountAddress,
+    entryPoint = getEntryPoint(chain, { version: "0.7.0" }),
+    isGlobalValidation = true,
+    entityId = 0n,
+  } = config;
+
+  if (entityId >= 2n ** 32n) {
+    throw new Error("Entity ID must be less than uint32.max");
+  }
+
+  const client = createBundlerClient({
+    transport,
+    chain,
+  });
+
+  const getAccountInitCode = async () => {
+    if (initCode) {
+      return initCode;
+    }
+
+    // If an initial owner is not provided, use the signer's address
+    const ownerAddress = initialOwner || (await signer.getAddress());
+
+    return concatHex([
+      factoryAddress,
+      encodeFunctionData({
+        abi: accountFactoryAbi,
+        functionName: "createAccount",
+        args: [ownerAddress, salt, DEFAULT_OWNER_ENTITY_ID],
+      }),
+    ]);
+  };
+
+  const _accountAddress = await getAccountAddress({
+    client,
+    entryPoint,
+    accountAddress,
+    getAccountInitCode,
+  });
+
+  const baseAccount = await toSmartContractAccount({
+    transport,
+    chain,
+    entryPoint,
+    accountAddress: _accountAddress,
+    source: `SMAV2Account`,
+    getAccountInitCode,
+    ...standardExecutor,
+    ...singleSignerMessageSigner(signer),
+  });
+
+  // TODO: add deferred action flag
+  const getAccountNonce = async (nonceKey?: bigint): Promise<bigint> => {
+    // uint32 entityId + (bytes1 options) makes a uint40
+    const nonceKeySuffix: bigint =
+      entityId * 256n + (isGlobalValidation ? 1n : 0n);
```

Comment:
Is this the only place that `isGlobalValidation` gets used? If so, then I don't think we need to override the `getAccountNonce` method. instead we should provide a utility to `buildAccountNonce(nonceKey, isGlobalValidation, isDeferredExecute)` and you can then use that to craft the `nonceKey` when building a UO

we could also leverage a custom middleware on clients as well since passing an override all the time is annoying

Reply:
discussed offline, current implementation is fine for now